src/server/Agent.js

```diff
@@ -89,19 +105,21 @@ export default class Agent extends SideAgent {
             this._originalResponse = transaction.response.clone();
         }
 
-        return this._runTransaction(transaction, userSocket);
+        return this._runTransaction(transaction, this._firstConnectionSocket);
     }
 
     async _stop() {
         if (this.server && this.server.address()) {
-            const promiseToClose = this.server.closeAsync(); // unconditionally
+            if (!this._serverClosed)
+                this._serverClosed = this.server.closeAsync(); // no accepted connections yet
```

Comment:
This `if` statement is a red flag, but the core problem may be elsewhere. The condition is not exactly "no accepted connections yet". The promise is _also_ nil if we inherited the socket from another server. And if we did inherit from S1, then how we can be sure that S1 did not call closeAsync()? This code appears to imply that a "keepConnections" server S1 could have called closeAsync() in the same spot when it was stopping...

Please double check how this all fits together. Perhaps we are breaking the keepConnections feature by calling closeAsync() so aggressively?

We may also need to adjust this._serverClosed description, to reflect the (unusual) fact that the promise does not even exist if the server has not accepted the first connection and has not been stopped yet. It is not a promise that external code can easily meaningfully wait for (if we expose it)!

Reply:
> In this case this.server is also nil - and we cannot get here

I missed that fact. Sorry. I think this fact resolves this change request.